### ISSUE TITLE
Meander fill

### ISSUE THREAD

**Issue by lexelby** [inkstitch/inkstitch#1142]
It would be cool if Ink/Stitch could do a meandering fill, like [this](https://www.pixiewillowpatterns.com/product/free-machine-embroidery-stipple-quilting-blocks/).  This is sometimes referred to as "stippling".

I haven't been able to find any descriptions of algorithms for this, but I have an idea of how to do it.

It involves building a grid of hexagons and finding a path in it that fills as much of the space as possible.

1. Start with a shape and a hex grid: 
![IMG_20210409_103107__01.jpg](https://user-images.githubusercontent.com/4446653/114200189-f3041180-9922-11eb-82ec-f8ed8bbb9c79.jpg)
2. Find a random-ish path between the start and end: 
![IMG_20210409_103232__01.jpg](https://user-images.githubusercontent.com/4446653/114200279-0c0cc280-9923-11eb-9f68-3c3f623907c2.jpg)
3. Find a hexagon that touches the path in only one place: 
![IMG_20210409_103306__01.jpg](https://user-images.githubusercontent.com/4446653/114200353-1e86fc00-9923-11eb-9374-33439f4fbf9e.jpg)
4. Erase the spot where the hexagon touches and replace it with the other way around the hexagon:
![IMG_20210409_103330__01.jpg](https://user-images.githubusercontent.com/4446653/114200436-352d5300-9923-11eb-86d8-4883ec87c6db.jpg)
5. Repeat 3 & 4 as much as possible: 
![IMG_20210409_104012__01.jpg](https://user-images.githubusercontent.com/4446653/114200492-424a4200-9923-11eb-833f-e9f69fe06969.jpg)
6. Smooth out the path to make it curvy:
![IMG_20210409_104401__01.jpg](https://user-images.githubusercontent.com/4446653/114200549-51c98b00-9923-11eb-9706-9af22e1b8c82.jpg)

It's not perfect -- there are some gaps that aren't filled in all the way.  It's also possible, if we're really unlucky, to get stuck and be unable to find any hexagons in step 3.  In that case we can allow finding a hexagon that touches on only 2 adjacent sides.

There's quite a bit of randomness involved here.  We could encode a random number generator seed as one of the Params. We could let the user specify the seed or press a "shuffle" button if they don't like the fill path they ended up with.  That may help with the situation in my example pictures where there are gaps.

I think this could be really neat!  I'm pretty sure users coming from commercial programs miss this feature.

**Comment by lexelby** [comment]
It might be possible to do this with a square grid too.

We could also allow the user to specify a smoothing factor.  Maybe they want a hexagonal fill!  scipy.interpolate has functions for smoothing paths.

**Comment by bastanja** [comment]
Hello, here is a paper that describes an algorithm to create stippling patterns:
https://archive.bridgesmathart.org/2015/bridges2015-231.pdf

**Comment by lexelby** [comment]
Wow!  Thank you so much!  I've looked for such papers and failed to find any.  Just the looking at the second page shows an algorithm that's really simple and easy to understand.

**Comment by lexelby** [comment]
At the simplest, we can build the s panning tree, convert it to shapely LineStrings, and then use `buffer()` to convert it to a path.  Neat!

One problem with this method is that it always ends where it starts.  Ink/Stitch users are used to being able to specify a start and end point.  I'm not sure if it's possible to adapt this algorithm to do that.

**Comment by tyrosinase** [comment]
On 4/10/21 12:07 PM, Lex Neva wrote:
>
> At the simplest, we can build the s panning tree, convert it to 
> shapely LineStrings, and then use |buffer()| to convert it to a path. 
> Neat!
>
> One problem with this method is that it always ends where it starts. 
> Ink/Stitch users are used to being able to specify a start and end 
> point. I'm not sure if it's possible to adapt this algorithm to do that.
>
Eh, if it ends where it starts then it's a loop that can be joined, then 
re-broken somewhere else (manually, if necessary). That's not too hard 
to live with.

**Comment by MK-12-20MM** [comment]
Just a comment .. This would be useful for cross-stitch emulation - some of the mainstream software supports this. I have not stitched one out, but - there are a few free ones that I have loaded into inkstitch and ran the simulator on to just observe the stitch pattern in the file. The ones I have looked at always start and stop at the same point.  It would not think that would be a problem and just a consideration of the fill type.   

![image](https://user-images.githubusercontent.com/10312448/114305056-14d4d400-9a9c-11eb-833f-43459e87a38f.png)

**Comment by lexelby** [comment]
Interesting, thank you!  For the second picture, does it go over every line twice?

**Comment by MK-12-20MM** [comment]
> Interesting, thank you! For the second picture, does it go over every line twice?

Appears the case  - This is where the free files are at ->    https://forum.embroideres.com/files/category/18-cross-stitch/

**Comment by lexelby** [comment]
Thank you!

**Comment by lexelby** [comment]
Actually, now that I think about it, the algorithm in the quilting paper can be modified only slightly to start and end wherever you please.

You calculate the spanning tree of the graph like in their algorithm. Then you find the shortest path from the start to the end in that spanning tree.  You follow that path, and at every node, if there's a branch available, you follow out and around it just as in the quilting algorithm.

That gets you right back where you started on the shortest path.  Then you continue on down the shortest path as before.

In other words, you traverse the shortest path once, and go around everything else twice.

I do wonder if that might make the shortest path visually obvious/distracting, but I think it would probably be well-disguised.

This probably works best for a hex-grid path.  It may be a little tricky to handle if two of the subpaths branch off from the shortest path at the same node.  Doable, though.

I guess I might be mostly talking to myself here, but I want to leave myself at least a bit of a clue for when I revisit this in 3 years ;)

**Comment by claudinepeyrat06** [comment]
Wondering wether  "space colonization algorithms" could not be useful ?
https://medium.com/@jason.webb/space-colonization-algorithm-in-javascript-6f683b743dc5
https://generativelandscapes.wordpress.com/2019/07/03/networks-and-space-colonization/

**Comment by lexelby** [comment]
Neat!  Unfortunately it looks like the or-tools path generator crosses holes (which we don't want), but Concorde is a GUI program that we can't integrate into Ink/Stitch.  I do think there's useful code in there that we may be able to incorporate into our solution, though.

**Comment by claudinepeyrat06** [comment]
@lexelby 
 If I understand well using Voronoi you get a set of nodes that you need to connect  using a path that does not cross itself and that stays in  the object to be filled  . 
You can achieve this  with  an ordering of the nodes, and a path from one node to the next
TSP is a way to order, but it is never going to give a fast algorithm anyway.
To order the nodes, why not use  a fill space curve like Hilbert curve?

If the shape to fill does not have holes, it should be sufficient.
If there are holes, then I think that there is an idea for a solution  here : 
https://www.researchgate.net/publication/319643051_Hilbert's_Space-filling_Curve_for_Regions_with_Holes
No code is provided, but they do give a recursive algorithm.


As a final user, I would love to get a loose fill and be able to use pattern on path to densify.

**Comment by lexelby** [comment]
Neat paper, thanks!

I would love to have hilbert curves, but I couldn't find anything useful that explained how to fill an arbitrary shape with a Hilbert curve, even without holes.  Have you found anything?

Another way to handle path generation is the technique described in the quilting paper, where you build a spanning tree and walk around the outside of it.  I bet that'd be really cool with a voronoi stippling.

**Comment by claudinepeyrat06** [comment]
Hi,
Yes walking around  a spaning tree of a voronoi stippling  would work.

Actually to fill an arbitrary full shape S  with an hilbert curve you only have to deform  the square that bounds the Hilbert curve to the shape of  S.  
 I  tried the following in inkscape:
- draw a  closed path P 
- draw an Hilbert curve  H (using the L-systems extension) that is roughtly the same "size"as P
- Turn the nodes of H  to smooth
- apply the "enveloppe deformation" path effect to H, mapping the square boundary of H to the shape of P

Of course the results are far from perfect , but it was fun to try
![hilbertasafilltrial](https://user-images.githubusercontent.com/88194877/131561484-1722b7fd-ee52-42a9-b5fc-b7b7e0428117.png)

**Comment by Tina-Chris** [comment]
That would be very neat indeed. Looking fowart to it :)!

**Comment by lexelby** [comment]
Yes, we can't use concorde because it is a GUI application, I think.  Also, it only runs on Linux and Windows, not Mac.  It looks like all of those links use concorde under the hood.

**Comment by lexelby** [comment]
Did it!
